Show a progress indicator when dragging and dropping PDF files to JabRef

BaronMac08 opened the pull request:
Closes [#8738 ](https://github.com/JabRef/jabref/issues/8738)

Builds upon: the branch `fix-melting-630` opens at [#12001 ](https://github.com/JabRef/jabref/pull/12001)

Modifications:
- Dragging and dropping several PDF files into JabRef will be one large import rather than one small import for every file

Additions:
- Added informative messages about the ongoing progress of the big import by building upon the BackgroundTask progress indicator

UI Screenshots:
![Screenshot 2024-10-20 at 11 57 57](https://github.com/user-attachments/assets/c3814e24-f285-4581-8bcb-f88368337ff0)
![Screenshot 2024-10-20 at 11 58 42](https://github.com/user-attachments/assets/07d110a1-e6bf-4ce5-8c77-ef634aa6dca1)
![Screenshot 2024-10-20 at 11 59 00](https://github.com/user-attachments/assets/d6976846-4550-4525-9128-fe6c866d3889)
![Screenshot 2024-10-20 at 11 59 47](https://github.com/user-attachments/assets/c4ec2e7b-e3a4-4a04-91db-85c4c1186a21)


### Mandatory checks

<!-- 
- Go through the list below. Please don't remove any items.
- [x] done; [ ] not done / not applicable
-->

- [x] I own the copyright of the code submitted and I licence it under the [MIT license](https://github.com/JabRef/jabref/blob/main/LICENSE)
- [x] Change in `CHANGELOG.md` described in a way that is understandable for the average user (if applicable)
- [ ] Tests created for changes (if applicable)
- [x] Manually tested changed features in running JabRef (always required)
- [x] Screenshots added in PR description (for UI changes)
- [x] [Checked developer's documentation](https://devdocs.jabref.org/): Is the information available and up to date? If not, I outlined it in this pull request.
- [x] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not, I created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, I submitted a pull request to the documentation repository.

InAnYan commented on src/main/java/org/jabref/gui/externalfiles/ImportHandler.java:
```diff
@@ -118,8 +125,10 @@ public List<ImportFilesResultItemViewModel> call() {
-                        updateMessage(Localization.lang("Processing file %0", file.getFileName()));
-                        updateProgress(counter, files.size() - 1d);
+                        setTitle(Localization.lang("Processing %0 file(s) for adding to %1.", files.size(), bibDatabaseContext.getDatabasePath().map(path -> path.getFileName().toString()).orElseGet(() -> "untitled")));
+                        updateMessage(Localization.lang("Processing file %0 | %1 of %2 file(s) processed.", file.getFileName(), counter, files.size()));
```
When the file name is too long, users have to scroll horizontaly in the task viewer.

It would be nice if you could introduce a function to reduce long strings into fixed size strings by cutting the end and replacing it with "...".

Other JabRef features could benefit from this function :)

koppor replied:
Be awre that the filename is the most important part and comes last. Should not be replaced by ...

Otherwise StringUtil.limitStringLength is the function to use.

BaronMac08 replied:
Thank you for the feedback. I will work on an implementation that prominently displays the filename because it is the most important part and modifies the message so that users do not have to scroll horizontally.

InAnYan commented:
I like it, what do you think Oliver?

koppor commented:
> I like it, what do you think Oliver?

To enable proper reviewing and handling, we need to get https://github.com/JabRef/jabref/pull/12001 in. Maybe, you can try it on Linux? (modifier keys do not work on macOS; thus Chris cannot properly test)

koppor commented:
I changed base to `fix-melting-630` to ease reviewing. According to https://stackoverflow.com/a/32658536/873282, this should work smoothly.

koppor commented on CHANGELOG.md:
```diff
@@ -40,6 +40,7 @@ Note that this project **does not** adhere to [Semantic Versioning](https://semv
 - We added automatic browser extension install on Windows for Chrome and Edge. [#6076](https://github.com/JabRef/jabref/issues/6076)
 - We added a search bar for filtering keyboard shortcuts. [#11686](https://github.com/JabRef/jabref/issues/11686)
 - By double clicking on a local citation in the Citation Relations Tab you can now jump the linked entry. [#11955](https://github.com/JabRef/jabref/pull/11955)
+- When dragging and dropping several PDF files into the Main Table, the BackgroundTask menu icon becomes a progress indicator visualising the import's progress. [#12035](https://github.com/JabRef/jabref/pull/12035)
```
Changelog entries are for the end user. No technical terms (also no technical casing). OMG, while reviewing I found, that "Background Task" appears in the UI, too. Needs to be fixed (I filed an internal issue https://github.com/JabRef/jabref-issue-melting-pot/issues/644).

```suggestion
- When dragging and dropping several PDF files into the main table, the menu icon for background tasks becomes a progress indicator visualising the import's progress. [#12035](https://github.com/JabRef/jabref/pull/12035)
```

koppor commented on src/main/java/org/jabref/gui/externalfiles/ImportHandler.java:
```diff
@@ -119,8 +121,13 @@ public List<ImportFilesResultItemViewModel> call() {
-                        updateProgress(counter, files.size() - 1d);
+                        setTitle(Localization.lang("Processing %0 file(s) for %1 | %2 of %0 file(s) processed.",
+                                files.size(),
+                                bibDatabaseContext.getDatabasePath().map(path -> path.getFileName().toString()).orElse("untitled"),
```
```suggestion
                                bibDatabaseContext.getDatabasePath().map(path -> path.getFileName().toString()).orElse(Localization.lang("untitled")),
```

koppor commented on src/main/java/org/jabref/model/strings/StringUtil.java:
```diff
@@ -167,6 +168,83 @@ public static String getCorrectFileName(String orgName, String defaultExtension)
+            return fileName;
+        }
+
+        // Reference the ellipsis' length for future usage
```
Remove this comment - this is obvios.

koppor commented on src/main/java/org/jabref/model/strings/StringUtil.java:
```diff
@@ -167,6 +168,83 @@ public static String getCorrectFileName(String orgName, String defaultExtension)
+        // Reference the ellipsis' length for future usage
+        final int ELLIPSIS_LENGTH = ELLIPSIS.length();
+
+        // Find the index of the extension dot
```
Remove this comment - this is obvios.

koppor commented on src/main/java/org/jabref/model/strings/StringUtil.java:
```diff
@@ -167,6 +168,83 @@ public static String getCorrectFileName(String orgName, String defaultExtension)
+        // Find the index of the extension dot
+        int extensionIndex = fileName.lastIndexOf('.');
+
+        // Extract the name and extension part of a file name
```
Remove this comment - this is obvios.

koppor commented on src/main/java/org/jabref/model/strings/StringUtil.java:
```diff
@@ -167,6 +168,83 @@ public static String getCorrectFileName(String orgName, String defaultExtension)
+            extension = "";
+        }
+
+        // Calculate the necessary length for the ELLIPSIS and extension
```
Remove this comment - this is obvios.

koppor commented on src/main/java/org/jabref/model/strings/StringUtil.java:
```diff
@@ -167,6 +168,83 @@ public static String getCorrectFileName(String orgName, String defaultExtension)
+        String name;
+        String extension;
+
+        // Check if file has an extension or not
```
Please refactor to use `org.jabref.logic.util.io.FileUtil#getFileExtension(...)`

It is OK for me to use `FileUtil.getFileExtension(fileName).map(extension -> "." + extension).orElse("")`.

koppor commented on src/test/java/org/jabref/model/strings/StringUtilTest.java:
```diff
@@ -367,6 +367,25 @@ void capitalizeFirst() {
         assertEquals("Aa", StringUtil.capitalizeFirst("AA"));
     }
 
+    @Test
+    void shortenFileName() {
+        assertEquals("", StringUtil.shortenFileName(null, 10));
+        assertEquals("", StringUtil.shortenFileName("thisisatestfile.pdf", 0));
+        assertEquals("", StringUtil.shortenFileName("example.txt", -5));
+        assertEquals("COMP101.pdf", StringUtil.shortenFileName("COMP101.pdf", 20));
+        assertEquals("COMP10101.pdf", StringUtil.shortenFileName("COMP10101.pdf", 13));
+        assertEquals("long...name.txt", StringUtil.shortenFileName("longfilename.txt", 15));
+        assertEquals("veryl...name", StringUtil.shortenFileName("verylongfilename", 12));
+        assertEquals("...", StringUtil.shortenFileName("COMP20202020.fbx", 3));
+        assertEquals("", StringUtil.shortenFileName("COMP20202020.txt", 2));
+        assertEquals("...", StringUtil.shortenFileName("COMP20202020.txt", 4));
+        assertEquals("long...ame.txt", StringUtil.shortenFileName("longfilename.txt", 14));
+        assertEquals("lo...me.txt", StringUtil.shortenFileName("long.file.name.txt", 11));
+        assertEquals("lo...me.extremelylongextension", StringUtil.shortenFileName("longfilename.extremelylongextension", 30));
+        assertEquals("fi...e.txt", StringUtil.shortenFileName("filename.txt", 10));
+        assertEquals("filename.txt", StringUtil.shortenFileName("filename.txt", 20));
+    }
```
Refactor to `@ParameterizedTest`. Use `@CsvSource`.

BaronMac08 commented:
Thank you for the feedback, @koppor. I have completed the refactoring as you suggested.

koppor commented on src/main/java/org/jabref/model/strings/StringUtil.java:
```diff
@@ -177,34 +178,29 @@ public static String getCorrectFileName(String orgName, String defaultExtension)
         String name;
         String extension;
 
         // Check if file has an extension or not
```
Remove this line, because the next line is the comment.

koppor commented on src/main/java/org/jabref/model/strings/StringUtil.java:
```diff
@@ -177,34 +178,29 @@ public static String getCorrectFileName(String orgName, String defaultExtension)
         }
 
-        // Calculate the necessary length for the ELLIPSIS and extension
         int totalNeededLength = ELLIPSIS_LENGTH + extension.length();
```
Move that constant to line 31 (so that it is a class constant)

koppor commented on src/main/java/org/jabref/model/strings/StringUtil.java:
```diff
@@ -177,34 +178,29 @@ public static String getCorrectFileName(String orgName, String defaultExtension)
             name = fileName;
-            extension = "";
+        } else {
+            name = fileName.substring(0, extensionIndex);
```
Do it differently - use `length() - extension.length()` - and remove line 191 (` int extensionIndex`) --> There should be only one way be followed to determine and handle extensions.

koppor commented:
@BaronMac08 Please also fix the unit test for `StringUtil`. Just adapt the number of allowed lines.

BaronMac08 commented:
@koppor I have refactored the comments and extension handling, modified the unit tests by choosing three representative file names, and changed the maxLength of each case to represent potential use cases.

Siedlerchr commented:
You need to re-open your PR and target the main branch!

BaronMac08 commented:
Hi @Siedlerchr, I have created a new PR to target the main branch at [#12068](https://github.com/JabRef/jabref/pull/12068). I see there are some conflicts between main and my branch. Would you like me to `git merge upstream/main` and resolve the conflicts locally?
